microphone/microphone_driver.c

```diff
@@ -0,0 +1,815 @@
+      goto error;
+
+   if (!string_is_empty(settings->arrays.microphone_resampler))
+      strlcpy(mic_st->resampler_ident,
+            settings->arrays.microphone_resampler,
+            sizeof(mic_st->resampler_ident));
+   else
+      mic_st->resampler_ident[0] = '\0';
+
+   mic_st->resampler_quality     = microphone_driver_get_resampler_quality(settings);
+   mic_st->source_ratio_original = (double)settings->uints.audio_output_sample_rate / settings->uints.microphone_sample_rate;
+   if (!isfinite(mic_st->source_ratio_original))
+   {
+      mic_st->source_ratio_original = 1.0f;
+   }
+   mic_st->source_ratio_current  = mic_st->source_ratio_original;
+
+   if (!retro_resampler_realloc(
+            &mic_st->resampler_data,
+            &mic_st->resampler,
+            mic_st->resampler_ident,
+            mic_st->resampler_quality,
+            mic_st->source_ratio_original))
+   {
+      RARCH_ERR("[Microphone]: Failed to initialize resampler \"%s\".\n", mic_st->resampler_ident);
+      goto error;
+   }
+
+   RARCH_LOG("[Microphone]: Initialized microphone driver\n");
+
+   /* The mic driver was initialized, now we're ready to open mics */
+   mic_st->flags |= MICROPHONE_DRIVER_FLAG_ACTIVE;
+
+   if (!microphone_driver_start(false))
+      goto error;
+
+   return true;
+
+error:
+   RARCH_ERR("[Microphone]: Failed to start microphone driver. Will continue without audio input.\n");
+   mic_st->flags &= ~MICROPHONE_DRIVER_FLAG_ACTIVE;
+   return microphone_driver_deinit();
+}
+
+/**
+ *
+ * @param microphone Handle to the microphone to init with a context
+ */
+static bool mic_driver_open_mic_internal(retro_microphone_t* microphone)
+{
+   microphone_driver_state_t *mic_st     = &mic_driver_st;
+   settings_t *settings                  = config_get_ptr();
+   const microphone_driver_t *mic_driver = mic_st->driver;
+   void *driver_context                  = mic_st->driver_context;
+   unsigned runloop_audio_latency        = runloop_state_get_ptr()->audio_latency;
+   unsigned setting_audio_latency        = settings->uints.microphone_latency;
+   unsigned actual_sample_rate           = 0;
+   unsigned audio_latency                = MAX(runloop_audio_latency, setting_audio_latency);
+   float slowmotion_ratio                = settings->floats.slowmotion_ratio;
+   size_t insamples_max                  = AUDIO_CHUNK_SIZE_NONBLOCKING * 1 * AUDIO_MAX_RATIO * slowmotion_ratio;
+
+   if (!microphone || !mic_driver || !(mic_st->flags & MICROPHONE_DRIVER_FLAG_ACTIVE))
+      return false;
+
+   microphone->sample_buffer_length = insamples_max * sizeof(int16_t);
+   microphone->sample_buffer        =
+         (int16_t*)memalign_alloc(64, microphone->sample_buffer_length);
+
+   if (!microphone->sample_buffer)
+      goto error;
+
+   microphone->microphone_context = mic_driver->open_mic(driver_context,
+      *settings->arrays.microphone_device ? settings->arrays.microphone_device : NULL,
+      settings->uints.microphone_sample_rate,
```

Comment:
Related to the other comment, this instance of `microphone_sample_rate` should probably be replaced by the proposed `core_input_sample_rate`. This way we are asking the driver to give as audio samples at the exact rate the core expects and in the best case we can skip going through the resampler. If the driver was unable to comply with our request, we'll set `microphone_sample_rate` to the actual sample rate below and the resampler ratio will be set to `core_input_sample_rate / microphone_sample_rate`.